### ISSUE TITLE
"panic: V" when analzing a project with type parameters

### ISSUE THREAD

**Issue by arp242** [dominikh/go-tools#1234]
I get various different panics when running this on a project type parameters (as just `staticcheck ./...`). This is with the current master (23e1086) (detailed stack traces below):

	panic: K
	panic: V
	panic: interface conversion: ast.Expr is *ast.IndexListExpr, not *ast.Ident
	panic: no concrete method: func (*zgo.at/zcache/v2.cache[K, V]).Add(k K, v V) error
	panic: no concrete method: func (*zgo.at/zcache/v2.cache[MainK, V]).Add(k MainK, v V) error

The different errors are presumably because things are run in parallel and different invocations hit a different error first; perhaps they're all the same underlying issue, but I didn't investigate.

I'm running this on https://github.com/arp242/zcache/tree/v2 – note this is on the "v2" branch; I'm running it on commit 73e1915 in that branch (the most recent at the time of writing).

I realize that support for this is still on-going work, but another issue mentioned it's fine to report crashes for this.

<details>
<summary>Stack traces</summary>
<pre>
panic: K

goroutine 172 [running]:
honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000f94000, {0x9c4af0?, 0xc0002197a0?}, 0x0)
        /data/src/go-tools/go/ir/methods.go:237 +0x5b1
honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000f94000, {0x9c49b0?, 0xc0005a51e0?}, 0x0)
        /data/src/go-tools/go/ir/methods.go:202 +0x387
honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000f94000, {0x9c4ac8?, 0xc000a32870?}, 0x0)
        /data/src/go-tools/go/ir/methods.go:233 +0x708
honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000f94000, {0x9c4a50?, 0xc0005ba1c0?}, 0x0)
        /data/src/go-tools/go/ir/methods.go:209 +0x448
honnef.co/go/tools/go/ir.(*Program).needMethodsOf(0xc000f94000, {0x9c4a50?, 0xc0005ba1c0?})
        /data/src/go-tools/go/ir/methods.go:145 +0x70
honnef.co/go/tools/go/ir.(*Package).build(0xc000448d80)
        /data/src/go-tools/go/ir/builder.go:2427 +0x111
sync.(*Once).doSlow(0xc000f94000?, 0xc000a85ae0?)
        /usr/lib/go/src/sync/once.go:68 +0xc2
sync.(*Once).Do(...)
        /usr/lib/go/src/sync/once.go:59
honnef.co/go/tools/go/ir.(*Package).Build(...)
        /data/src/go-tools/go/ir/builder.go:2415
honnef.co/go/tools/internal/passes/buildir.run(0xc000f88270)
        /data/src/go-tools/internal/passes/buildir/buildir.go:86 +0x368
honnef.co/go/tools/lintcmd/runner.(*analyzerRunner).do(0xc000af19e0, {0x9c72b8?, 0xc000c503c0})
        /data/src/go-tools/lintcmd/runner/runner.go:1032 +0x614
honnef.co/go/tools/lintcmd/runner.genericHandle({0x9c72b8?, 0xc000c503c0}, {0x9c72b8?, 0xc000c50140?}, 0x1?, 0x0, 0xc000c3ae40)
        /data/src/go-tools/lintcmd/runner/runner.go:857 +0x102
honnef.co/go/tools/lintcmd/runner.(*subrunner).runAnalyzers(0xc0000b2730, 0xc00057ad80, 0xc000a85b30)
        /data/src/go-tools/lintcmd/runner/runner.go:1105 +0x766
honnef.co/go/tools/lintcmd/runner.(*subrunner).doUncached(0xc0000b2730, 0xc00057ad80)
        /data/src/go-tools/lintcmd/runner/runner.go:728 +0x1c9
honnef.co/go/tools/lintcmd/runner.(*subrunner).do(0xc0000b2730, {0x9c7300?, 0xc00057ad80})
        /data/src/go-tools/lintcmd/runner/runner.go:584 +0x78d
honnef.co/go/tools/lintcmd/runner.(*Runner).Run.func2({0x9c7300?, 0xc00057ad80?})
        /data/src/go-tools/lintcmd/runner/runner.go:1308 +0x30
honnef.co/go/tools/lintcmd/runner.genericHandle({0x9c7300?, 0xc00057ad80}, {0x9c7300?, 0xc00057ac60?}, 0xc0000547d0?, 0xc0000f7c60, 0xc00041e110)
        /data/src/go-tools/lintcmd/runner/runner.go:857 +0x102
created by honnef.co/go/tools/lintcmd/runner.(*Runner).Run
        /data/src/go-tools/lintcmd/runner/runner.go:1307 +0x525
</pre>

<pre>
panic: V

goroutine 159 [running]:
honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000fa2000, {0x9c4af0?, 0xc0009bdf80?}, 0x0)
        /data/src/go-tools/go/ir/methods.go:237 +0x5b1
honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000fa2000, {0x9c4aa0?, 0xc0009d2000?}, 0x1)
        /data/src/go-tools/go/ir/methods.go:228 +0x669
honnef.co/go/tools/go/ir.(*Program).needMethods(0xc000fa2000, {0x9c49d8?, 0xc0008107e0?}, 0x0)
        /data/src/go-tools/go/ir/methods.go:221 +0x588
honnef.co/go/tools/go/ir.(*Program).needMethodsOf(0xc000fa2000, {0x9c49d8?, 0xc0008107e0?})
        /data/src/go-tools/go/ir/methods.go:145 +0x70
honnef.co/go/tools/go/ir.(*Package).build(0xc0005b0d80)
        /data/src/go-tools/go/ir/builder.go:2427 +0x111
sync.(*Once).doSlow(0xc000fa2000?, 0xc0001a5c70?)
        /usr/lib/go/src/sync/once.go:68 +0xc2
sync.(*Once).Do(...)
        /usr/lib/go/src/sync/once.go:59
honnef.co/go/tools/go/ir.(*Package).Build(...)
        /data/src/go-tools/go/ir/builder.go:2415
honnef.co/go/tools/internal/passes/buildir.run(0xc000673e10)
        /data/src/go-tools/internal/passes/buildir/buildir.go:86 +0x368
honnef.co/go/tools/lintcmd/runner.(*analyzerRunner).do(0xc000eaa3c0, {0x9c72b8?, 0xc000c41400})
        /data/src/go-tools/lintcmd/runner/runner.go:1032 +0x614
honnef.co/go/tools/lintcmd/runner.genericHandle({0x9c72b8?, 0xc000c41400}, {0x9c72b8?, 0xc000c412c0?}, 0x0?, 0xc0003bea80, 0xc000b94330)
        /data/src/go-tools/lintcmd/runner/runner.go:857 +0x102
created by honnef.co/go/tools/lintcmd/runner.(*subrunner).runAnalyzers
        /data/src/go-tools/lintcmd/runner/runner.go:1101 +0x705
</pre>

<pre>
panic: interface conversion: ast.Expr is *ast.IndexListExpr, not *ast.Ident

goroutine 227 [running]:
honnef.co/go/tools/stylecheck.CheckExportedFunctionDocs.func1({0x9c4410?, 0xc00019df80?})
        /data/src/go-tools/stylecheck/lint.go:809 +0x385
golang.org/x/tools/go/ast/inspector.(*Inspector).Preorder(0xc000d6e948, {0xc00025ae00?, 0x1?, 0x8e48c0?}, 0xc000654df0)
        /home/martin/.cache/go/pkg/mod/golang.org/x/tools@v0.1.8/go/ast/inspector/inspector.go:77 +0x9a
honnef.co/go/tools/analysis/code.Preorder(...)
        /data/src/go-tools/analysis/code/visit.go:16
honnef.co/go/tools/stylecheck.CheckExportedFunctionDocs(0x8838a0?)
        /data/src/go-tools/stylecheck/lint.go:826 +0x8f
honnef.co/go/tools/lintcmd/runner.(*analyzerRunner).do(0xc0008744e0, {0x9c72b8?, 0xc000bf8460})
        /data/src/go-tools/lintcmd/runner/runner.go:1032 +0x614
honnef.co/go/tools/lintcmd/runner.genericHandle({0x9c72b8?, 0xc000bf8460}, {0x9c72b8?, 0xc000229ea0?}, 0x0?, 0xc0004a7030, 0xc000b88c90)
        /data/src/go-tools/lintcmd/runner/runner.go:857 +0x102
created by honnef.co/go/tools/lintcmd/runner.(*subrunner).runAnalyzers
        /data/src/go-tools/lintcmd/runner/runner.go:1101 +0x705
</pre>

<pre>
panic: no concrete method: func (*zgo.at/zcache/v2.cache[K, V]).Add(k K, v V) error

goroutine 171 [running]:
honnef.co/go/tools/go/ir.(*Program).declaredFunc(0xc0000d0000, 0xc000f11e50)
        /data/src/go-tools/go/ir/methods.go:124 +0xf9
honnef.co/go/tools/go/ir.makeWrapper(0xc0000d0000, 0xc000f11ea0)
        /data/src/go-tools/go/ir/wrappers.go:125 +0xf9b
honnef.co/go/tools/go/ir.(*Program).addMethod(0x9c49d8?, 0xc0007c10d0, 0xc000f11ea0)
        /data/src/go-tools/go/ir/methods.go:84 +0x139
honnef.co/go/tools/go/ir.(*Program).needMethods(0xc0000d0000, {0x9c49d8?, 0xc000317f10?}, 0x0)
        /data/src/go-tools/go/ir/methods.go:173 +0x785
honnef.co/go/tools/go/ir.(*Program).needMethodsOf(0xc0000d0000, {0x9c49d8?, 0xc000317f10?})
        /data/src/go-tools/go/ir/methods.go:145 +0x70
honnef.co/go/tools/go/ir.(*Package).build(0xc0001a8510)
        /data/src/go-tools/go/ir/builder.go:2427 +0x111
sync.(*Once).doSlow(0xc0000d0000?, 0xc000829d10?)
        /usr/lib/go/src/sync/once.go:68 +0xc2
sync.(*Once).Do(...)
        /usr/lib/go/src/sync/once.go:59
honnef.co/go/tools/go/ir.(*Package).Build(...)
        /data/src/go-tools/go/ir/builder.go:2415
honnef.co/go/tools/internal/passes/buildir.run(0xc0007bc410)
        /data/src/go-tools/internal/passes/buildir/buildir.go:86 +0x368
honnef.co/go/tools/lintcmd/runner.(*analyzerRunner).do(0xc000b7ea80, {0x9c72b8?, 0xc000b54320})
        /data/src/go-tools/lintcmd/runner/runner.go:1032 +0x614
honnef.co/go/tools/lintcmd/runner.genericHandle({0x9c72b8?, 0xc000b54320}, {0x9c72b8?, 0xc000b541e0?}, 0x0?, 0xc00041b440, 0xc000b7c6b0)
        /data/src/go-tools/lintcmd/runner/runner.go:857 +0x102
created by honnef.co/go/tools/lintcmd/runner.(*subrunner).runAnalyzers
        /data/src/go-tools/lintcmd/runner/runner.go:1101 +0x705
</pre>

<pre>
panic: no concrete method: func (*zgo.at/zcache/v2.cache[MainK, V]).Add(k MainK, v V) error

goroutine 237 [running]:
honnef.co/go/tools/go/ir.(*Program).declaredFunc(0xc0003e0000, 0xc000cc9220)
        /data/src/go-tools/go/ir/methods.go:124 +0xf9
honnef.co/go/tools/go/ir.makeWrapper(0xc0003e0000, 0xc0010a76d0)
        /data/src/go-tools/go/ir/wrappers.go:125 +0xf9b
honnef.co/go/tools/go/ir.(*Program).addMethod(0x9c4a28?, 0xc000caef60, 0xc0010a76d0)
        /data/src/go-tools/go/ir/methods.go:84 +0x139
honnef.co/go/tools/go/ir.(*Program).needMethods(0xc0003e0000, {0x9c4a28?, 0xc0008ce3e0?}, 0x0)
        /data/src/go-tools/go/ir/methods.go:173 +0x785
honnef.co/go/tools/go/ir.(*Program).needMethods(0xc0003e0000, {0x9c4ac8?, 0xc0008c88e8?}, 0x0)
        /data/src/go-tools/go/ir/methods.go:233 +0x708
honnef.co/go/tools/go/ir.(*Program).needMethods(0xc0003e0000, {0x9c4a28?, 0xc000caea20?}, 0x0)
        /data/src/go-tools/go/ir/methods.go:182 +0x1ce
honnef.co/go/tools/go/ir.(*Program).needMethods(0xc0003e0000, {0x9c49d8?, 0xc0006aa620?}, 0x0)
        /data/src/go-tools/go/ir/methods.go:215 +0x55f
honnef.co/go/tools/go/ir.(*Program).needMethodsOf(0xc0003e0000, {0x9c49d8?, 0xc0006aa620?})
        /data/src/go-tools/go/ir/methods.go:145 +0x70
honnef.co/go/tools/go/ir.(*Package).build(0xc0000becf0)
        /data/src/go-tools/go/ir/builder.go:2427 +0x111
sync.(*Once).doSlow(0xc0003e0000?, 0xc00091d040?)
        /usr/lib/go/src/sync/once.go:68 +0xc2
sync.(*Once).Do(...)
        /usr/lib/go/src/sync/once.go:59
honnef.co/go/tools/go/ir.(*Package).Build(...)
        /data/src/go-tools/go/ir/builder.go:2415
honnef.co/go/tools/internal/passes/buildir.run(0xc0000a25b0)
        /data/src/go-tools/internal/passes/buildir/buildir.go:86 +0x368
honnef.co/go/tools/lintcmd/runner.(*analyzerRunner).do(0xc000b9b920, {0x9c72b8?, 0xc000c7d5e0})
        /data/src/go-tools/lintcmd/runner/runner.go:1032 +0x614
honnef.co/go/tools/lintcmd/runner.genericHandle({0x9c72b8?, 0xc000c7d5e0}, {0x9c72b8?, 0xc000c7d4a0?}, 0x507e54?, 0x0, 0xc000c5ee40)
        /data/src/go-tools/lintcmd/runner/runner.go:857 +0x102
honnef.co/go/tools/lintcmd/runner.(*subrunner).runAnalyzers(0xc0001ac730, 0xc0000c0ea0, 0xc00091d090)
        /data/src/go-tools/lintcmd/runner/runner.go:1105 +0x766
honnef.co/go/tools/lintcmd/runner.(*subrunner).doUncached(0xc0001ac730, 0xc0000c0ea0)
        /data/src/go-tools/lintcmd/runner/runner.go:728 +0x1c9
honnef.co/go/tools/lintcmd/runner.(*subrunner).do(0xc0001ac730, {0x9c7300?, 0xc0000c0ea0})
        /data/src/go-tools/lintcmd/runner/runner.go:584 +0x78d
honnef.co/go/tools/lintcmd/runner.(*Runner).Run.func2({0x9c7300?, 0xc0000c0ea0?})
        /data/src/go-tools/lintcmd/runner/runner.go:1308 +0x30
honnef.co/go/tools/lintcmd/runner.genericHandle({0x9c7300?, 0xc0000c0ea0}, {0x9c7300?, 0xc0000c0d80?}, 0x0?, 0xc0003c0a80, 0xc0001dd010)
        /data/src/go-tools/lintcmd/runner/runner.go:857 +0x102
created by honnef.co/go/tools/lintcmd/runner.(*Runner).Run
        /data/src/go-tools/lintcmd/runner/runner.go:1307 +0x525
</pre>

</details>

<details>
<summary>Environment info etc.</summary>

<pre>
% staticcheck -debug.version
staticcheck (no version)

Compiled with Go version: go1.18
Main module:
        honnef.co/go/tools
Dependencies:
        github.com/BurntSushi/toml@v0.4.1 (sum: h1:GaI7EiDXDRfa8VshkTj7Fym7ha+y8/XxIgD2okUIjLw=)
        golang.org/x/crypto@v0.0.0-20220321153916-2c7772ba3064 (sum: h1:S25/rfnfsMVgORT4/J61MJ7rdyseOZOyvLIrZEZ7s6s=)
        golang.org/x/exp/typeparams@v0.0.0-20220218215828-6cf2b201936e (sum: h1:qyrTQ++p1afMkO4DPEeLGq/3oTsdlvdH4vqZUBWzUKM=)
        golang.org/x/mod@v0.5.1 (sum: h1:OJxoQ/rynoF0dcCdI7cLPktw/hR2cueqYfjm43oqK38=)
        golang.org/x/net@v0.0.0-20211112202133-69e39bad7dc2 (sum: h1:CIJ76btIcR3eFI5EgSo6k1qKw9KJexJuRLI9G7Hp5wE=)
        golang.org/x/sys@v0.0.0-20211019181941-9d821ace8654 (sum: h1:id054HUawV2/6IGm2IV8KZQjqtwAOo2CYlOToYqa0d0=)
        golang.org/x/text@v0.3.7 (sum: h1:olpwvP2KacW1ZWvsR7uQhoyTYvKAupfQrRGBFM352Gk=)
        golang.org/x/tools@v0.1.8 (sum: h1:P1HhGGuLW4aAclzjtmJdf0mJOjVUZUzOTqkAkWL+l6w=)
        golang.org/x/xerrors@v0.0.0-20200804184101-5ec99f83aff1 (sum: h1:go1bK/D/BFZV2I8cIQd1NKEZ+0owSTG1fDTci4IqFcE=)

% go version
go version go1.18 linux/amd64

% go env
GO111MODULE=""
GOARCH="amd64"
GOBIN="/home/martin/.local/gobin"
GOCACHE="/home/martin/.cache/go/cache"
GOENV="/home/martin/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS="-modcacherw"
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/martin/.cache/go/pkg/mod"
GONOPROXY="github.com/timescale/*"
GONOSUMDB="github.com/timescale/*"
GOOS="linux"
GOPATH="/home/martin/.cache/go"
GOPRIVATE="github.com/timescale/*"
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/lib/go"
GOSUMDB="sum.golang.org"
GOTMPDIR="/home/martin/.cache/go/tmp"
GOTOOLDIR="/usr/lib/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.18"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="0"
GOMOD="/home/martin/code/Golib/zcache/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/home/martin/.cache/go/tmp/go-build1975525597=/tmp/go-build -gno-record-gcc-switches"
</pre>
</details>
